x-pack/plugin/logstash/src/main/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineAction.java

```diff
@@ -122,37 +141,66 @@ private void handleFailure(Exception e, ActionListener<GetPipelineResponse> list
         }
     }
 
-    private void handleSearchResponse(
+    private void handleFilteringSearchResponse(
         SearchResponse searchResponse,
         Map<String, BytesReference> pipelineSources,
+        Set<String> explicitPipelineIds,
+        Set<String> wildcardPipelineIdPatterns,
+        int maxNumHitsToFetch,
         Consumer<SearchResponse> clearScroll,
         ActionListener<GetPipelineResponse> listener
     ) {
+        maxNumHitsToFetch += searchResponse.getHits().getHits().length;
         for (SearchHit hit : searchResponse.getHits().getHits()) {
-            pipelineSources.put(hit.getId(), hit.getSourceRef());
+            if (explicitPipelineIds.isEmpty() && wildcardPipelineIdPatterns.isEmpty()) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+
+            // take if ID is in request IDs set
+            if (explicitPipelineIds.contains(hit.getId())) {
+                pipelineSources.put(hit.getId(), hit.getSourceRef());
+                continue;
+            }
+            // take if id matches request wildcard pattern
+            for (String wildcardPipelineIdPattern : wildcardPipelineIdPatterns) {
+                if (Pattern.matches(wildcardPipelineIdPattern, hit.getId())) {
+                    pipelineSources.put(hit.getId(), hit.getSourceRef());
+                    break;
+                }
+            }
         }
 
-        if (pipelineSources.size() > searchResponse.getHits().getTotalHits().value) {
+        if (maxNumHitsToFetch > searchResponse.getHits().getTotalHits().value) {
             clearScroll.accept(searchResponse);
             listener.onFailure(
                 new IllegalStateException(
                     "scrolling returned more hits ["
-                        + pipelineSources.size()
+                        + maxNumHitsToFetch
                         + "] than expected ["
                         + searchResponse.getHits().getTotalHits().value
                         + "] so bailing out to prevent unbounded "
                         + "memory consumption."
                 )
             );
-        } else if (pipelineSources.size() == searchResponse.getHits().getTotalHits().value) {
+        } else if (maxNumHitsToFetch == searchResponse.getHits().getTotalHits().value) {
             clearScroll.accept(searchResponse);
             listener.onResponse(new GetPipelineResponse(pipelineSources));
         } else {
+            int finalNumHitsToFetch = maxNumHitsToFetch;
```

Comment:
Variable name is a little confusing here, can we make this something like "numberOfHitsSeenSoFar"? Then the method parameter could be something like "numberOfHitsSeenPreviously"?

The name "maxNumHitsToFetch" makes me think this is a hard limit on the results, but in reality it's just a counter and the hard limit is the scroll total hits.

Reply:
From paging point of view, this makes sense.